src/Features/LanguageServer/Protocol/Handler/MapCode/MapCodeHandler.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MapCode;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(MapCodeHandler)), Shared]
+[Method(WorkspaceMapCodeName)]
+internal sealed partial class MapCodeHandler : ILspServiceRequestHandler<MapCodeParams, LSP.WorkspaceEdit?>
+{
+    public const string WorkspaceMapCodeName = "workspace/mapCode";
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public MapCodeHandler()
+    {
+    }
+
+    public bool MutatesSolutionState => false;
+    public bool RequiresLSPSolution => true;
+
+    public async Task<WorkspaceEdit?> HandleRequestAsync(MapCodeParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Solution);
+
+        //TODO: handle request.Updates if not empty
+        if (request.Updates is not null)
+        {
+            context.TraceWarning("mapCode Request failed: additional workspace 'Update' is currently not supported");
+            return null;
+        }
+
+        using var _ = PooledDictionary<Uri, TextEdit[]>.GetInstance(out var uriToEditsMap);
+        foreach (var codeMapping in request.Mappings)
+        {
+            var mappingResult = await MapCodeAsync(codeMapping).ConfigureAwait(false);
+
+            if (mappingResult is not (Uri uri, TextEdit[] textEdits))
+            {
+                // Failed the entire request if any of the sub-requests failed
+                context.TraceWarning("mapCode Request failed: a sub-request failed");
+                return null;
+            }
+
+            if (!uriToEditsMap.TryAdd(uri, textEdits))
+            {
+                context.TraceWarning($"mapCode sub-request for {uri} failed: multiple MapCodeMappings for the same document is not supported.");
+                return null;
```

Comment:
I'm OK with this - but definitely consider throwing if this is entirely unexpected as then we'll get automated NFW reports.  If you just return null we won't necessarily know anything is going wrong from telemetry. 

But if its an expected as part of normal execution, then returning null is correct.

Same applies to the other locations you return null at.